File: rust/src/compact_dp.rs
use serde_json::Value;

/// DP compact decision configuration.
/// Matches DP_* env vars in bash-agent.
#[derive(Debug, Clone)]
pub struct DPCompactConfig {
    pub p_input: f64,
    pub p_cache: f64,
    pub p_out: f64,
    pub v: usize,
    pub s: usize,
    pub l_fixed: f64,
    pub baseline_e: i32,
    pub e_fixed: i32,
    pub r: f64,
    pub beta: f64,
    pub min_keep_ratio: f64,
}

impl Default for DPCompactConfig {
    fn default() -> Self {
        Self {
            p_input: 3.0,
            p_cache: 0.30,
            p_out: 15.0,
            v: 5000,
            s: 500,
            l_fixed: 0.0,
            baseline_e: 8,
            e_fixed: 0,
            r: 0.8,
            beta: 0.03,
            min_keep_ratio: 0.12,
        }
    }
}

/// Compute the optimal number of lines to keep using DP analysis.
/// All computation (E, L, avg) happens here — callers pass raw stats only.
/// prev_compactions: number of previous compactions (from stats).
/// current_turn: completed user-input rounds (from stats).
/// total_requests: total agent_request_count (from stats).
/// total_input_tokens: total cumulative input tokens (from stats).
/// Returns: Some(keep_lines) turn-aligned, or None if no compact beneficial.
pub fn compact_dp_decision(
    lines: &[Value],
    cfg: &DPCompactConfig,
    prev_compactions: usize,
    current_turn: usize,
    total_requests: usize,
    total_input_tokens: usize,
) -> Option<usize> {
    if lines.is_empty() {
        return None;
    }

    let n = lines.len();

    let mut sizes = Vec::with_capacity(n);
    let mut role_user = Vec::with_capacity(n);
    for line in lines {
        let s = serde_json::to_string(line).unwrap_or_default();
        sizes.push((s.len() + 3) / 4);
        let is_user = line.get("role").and_then(Value::as_str) == Some("user")
            && line.get("content").is_some_and(|c| c.is_string());
        role_user.push(is_user);
    }

    let total_tokens: usize = sizes.iter().sum();

    // E: expected remaining user-input rounds
    let e = if cfg.e_fixed > 0 {
        cfg.e_fixed as f64
    } else if cfg.baseline_e > 0 {
        let remaining = cfg.baseline_e - current_turn as i32;
        let floor = if cfg.baseline_e > 1 {
            cfg.baseline_e / 2
        } else {
            2
        };
        remaining.max(floor) as f64
    } else {
        2.0
    };

    // L: avg LLM calls per user input (auto from stats if l_fixed=0)
    let l = if cfg.l_fixed > 0.0 {
        cfg.l_fixed
    } else if current_turn > 0 && total_requests > 0 {
        let computed = total_requests as f64 / current_turn as f64;
        if computed >= 1.0 { computed } else { 5.0 }
    } else {
        5.0
    };
    let l = l.max(1.0);

    // avg: avg input tokens per LLM request (auto from stats)
    let avg = if total_requests > 0 && total_input_tokens > 0 {
        total_input_tokens as f64 / total_requests as f64
    } else {
        4000.0
    };

    // R = E * L
    let r_total = e * l;

    // r_t = r^(c+1) (independent of k)
    let mut r_t = cfg.r.powi((prev_compactions + 1) as i32);
    if r_t < 0.37 {
        r_t = 0.37;
    }

    let n_remain = r_total * avg;
    let info_loss = cfg.beta * (1.0 - r_t) * n_remain * cfg.p_input / 1_000_000.0;

    let l_instr = 70.0_f64;

    let min_keep = {
        let k = (n as f64 * cfg.min_keep_ratio + 0.5) as usize;
        k.max(3).min(n)
    };

    let mut best_k = 0usize;
    let mut best_benefit = f64::NEG_INFINITY;

    for k in min_keep..=n {
        let k_tokens: usize = sizes[n - k..].iter().sum();
        let h = total_tokens as f64 - k_tokens as f64;
        if h <= 0.0 {
            continue;
        }

        let kf = k_tokens as f64;
        let sf = cfg.s as f64;
        let vf = cfg.v as f64;

        // ① Savings: (R-1) * P_cache * H
        let savings = (r_total - 1.0) * cfg.p_cache * h / 1_000_000.0;

        // ② Cache invalidation: (S + K) * (P_base - P_cache)
        let cache_miss = (sf + kf) * (cfg.p_input - cfg.p_cache) / 1_000_000.0;

        // ③ Compaction request cost
        let compact_cost =
            (cfg.p_cache * (vf + h) + cfg.p_input * l_instr + cfg.p_out * sf) / 1_000_000.0;

        let benefit = savings - cache_miss - compact_cost - info_loss;

        if benefit > best_benefit {
            best_benefit = benefit;
            best_k = k;
        }
    }

    if best_benefit <= 0.0 {
        return None;
    }

    let mut adj = best_k;
    let mut cut = n - adj;
    while cut > 0 && !role_user[cut] {
        cut -= 1;
        adj = n - cut;
    }
    if adj < 1 {
        adj = 1;
    }
    Some(adj)
}

File: go/internal/conversation/compact_dp.go
package conversation

import (
	"encoding/json"
	"math"
	"strings"
)

// DPCompactConfig holds the parameters for the DP compact decision.
// Matches DP_* env vars in bash-agent.
type DPCompactConfig struct {
	PInput        float64 // $/MTok, uncached input price
	PCache       float64 // $/MTok, cached input price
	POut         float64 // $/MTok, output price
	V            int     // fixed prefix tokens (system prompt + tools + summary)
	S            int     // fixed summary length (tokens)
	LFixed       float64 // fixed L override (0 = auto from stats)
	BaselineE    int     // baseline for E calculation
	EFixed       int     // fixed E override (0 = use BaselineE)
	R            float64 // single-step summary retention rate
	Beta         float64 // info loss penalty coefficient
	MinKeepRatio float64 // minimum fraction of messages to retain
}

func DefaultDPCompactConfig() DPCompactConfig {
	return DPCompactConfig{
		PInput:        3.0,
		PCache:       0.30,
		POut:         15.0,
		V:            5000,
		S:            500,
		LFixed:       0,
		BaselineE:    8,
		EFixed:       0,
		R:            0.8,
		Beta:         0.03,
		MinKeepRatio: 0.12,
	}
}

// CompactTurnKeep computes turn-aligned lines to keep using MinKeepRatio.
// Matches bash compact_turn_keep(): counts user inputs, applies ratio, aligns to user boundary.
func (s Store) CompactTurnKeep(minKeepRatio float64) (int, error) {
	lines, err := s.Lines()
	if err != nil {
		return 0, err
	}
	if len(lines) == 0 {
		return 0, nil
	}

	// Detect user role: match {"role":"user","content":"... (excludes tool_result lines)
	roleUser := make([]bool, len(lines))
	totalTurns := 0
	for i, line := range lines {
		s := string(line)
		if strings.HasPrefix(s, `{"role":"user","content":"`) {
			roleUser[i] = true
			totalTurns++
		}
	}

	target := int(float64(totalTurns)*minKeepRatio + 0.5)
	if target < 1 {
		target = 1
	}
	if target > totalTurns {
		target = totalTurns
	}

	keep := 0
	found := 0
	for i := len(lines) - 1; i >= 0 && found < target; i-- {
		keep++
		if roleUser[i] {
			found++
		}
	}
	if keep < 3 {
		return len(lines), nil
	}
	return keep, nil
}

// CompactDPDecision computes the optimal number of lines to keep.
// All computation (E, L, avg) happens here — callers pass raw stats only.
// prevCompactions: number of previous compactions (from stats).
// currentTurn: completed user-input rounds (from stats).
// totalRequests: total LLM request count (from stats).
// totalInputTokens: total cumulative input tokens (from stats).
// Returns: keepLines (turn-aligned, or 0 if no compact beneficial).
func (s Store) CompactDPDecision(cfg DPCompactConfig, prevCompactions int, currentTurn int, totalRequests int, totalInputTokens int) (int, error) {
	lines, err := s.Lines()
	if err != nil {
		return 0, err
	}
	if len(lines) == 0 {
		return 0, nil
	}

	n := len(lines)
	sizes := make([]int, n)
	roleUser := make([]bool, n)
	for i, line := range lines {
		sizes[i] = (len(line) + 3) / 4
		if len(line) > 0 {
			var msg struct {
				Role    string          `json:"role"`
				Content json.RawMessage `json:"content"`
			}
			if err := json.Unmarshal(line, &msg); err == nil {
				roleUser[i] = msg.Role == "user" && len(msg.Content) > 0 && msg.Content[0] == '"'
			}
		}
	}

	totalTokens := 0
	for _, s := range sizes {
		totalTokens += s
	}

	// E: expected remaining user-input rounds
	var E float64
	if cfg.EFixed > 0 {
		E = float64(cfg.EFixed)
	} else if cfg.BaselineE > 0 {
		remaining := cfg.BaselineE - currentTurn
		floor := cfg.BaselineE / 2
		if floor < 2 {
			floor = 2
		}
		if remaining < floor {
			remaining = floor
		}
		E = float64(remaining)
	} else {
		E = 2.0
	}

	// L: avg LLM calls per user input (auto from stats if LFixed=0)
	var L float64
	if cfg.LFixed > 0 {
		L = cfg.LFixed
	} else if currentTurn > 0 && totalRequests > 0 {
		L = float64(totalRequests) / float64(currentTurn)
	} else {
		L = 5.0
	}
	if L < 1 {
		L = 1
	}

	// avg: avg input tokens per LLM request (auto from stats)
	var avg float64
	if totalRequests > 0 && totalInputTokens > 0 {
		avg = float64(totalInputTokens) / float64(totalRequests)
	} else {
		avg = 4000
	}

	// R = E * L: total expected remaining LLM calls
	R := E * L

	// Cumulative retention: r^(c+1) (independent of k)
	rT := math.Pow(cfg.R, float64(prevCompactions+1))
	if rT < 0.37 {
		rT = 0.37
	}

	// N_remain: expected remaining input tokens (R * avg_per_request)
	NRemain := R * avg

	// ④ Info loss (constant across all k)
	infoLoss := cfg.Beta * (1.0 - rT) * NRemain * cfg.PInput / 1_000_000

	// Summary instruction length (fixed ~70 tokens)
	lInstr := 70.0

	// Minimum keep (hard floor at 3)
	minKeep := int(float64(n)*cfg.MinKeepRatio + 0.5)
	if minKeep < 3 {
		minKeep = 3
	}
	if minKeep > n {
		minKeep = n
	}

	bestK := 0
	bestBenefit := math.Inf(-1)

	for k := minKeep; k <= n; k++ {
		K := 0
		for i := n - k; i < n; i++ {
			K += sizes[i]
		}

		H := totalTokens - K
		if H <= 0 {
			continue
		}

		Hf := float64(H)
		Kf := float64(K)
		Sf := float64(cfg.S)
		Vf := float64(cfg.V)

		// ① Savings: (R-1) subsequent LLM calls save P_cache × H each
		savings := (R - 1) * cfg.PCache * Hf / 1_000_000

		// ② Cache invalidation: (S + K) at P_base instead of P_cache
		cacheMiss := (Sf + Kf) * (cfg.PInput - cfg.PCache) / 1_000_000

		// ③ Compaction request cost
		compactCost := (cfg.PCache*(Vf+Hf) + cfg.PInput*lInstr + cfg.POut*Sf) / 1_000_000

		benefit := savings - cacheMiss - compactCost - infoLoss

		if benefit > bestBenefit {
			bestBenefit = benefit
			bestK = k
		}
	}

	if bestBenefit <= 0 {
		return 0, nil
	}

	// Align to user-message (turn) boundary
	adj := bestK
	cut := n - adj
	for cut > 0 && !roleUser[cut] {
		cut--
		adj = n - cut
	}
	if adj < 1 {
		adj = 1
	}
	return adj, nil
}
